### PR TITLE
Issue#34 camera switcher

### PR DESCRIPTION
**Pull request by UppaJung**

### CODE REVIEW

**Review by UppaJung — COMMENTED** [review]
I'll try to get this to run locally.  For now based on reviewing the code...

Issues to fix:

1. The set of cameras is determined once and never updated, so if I connect or disconnect a camera the picker will not get the updated list. Nor will it get updated if I open the screen of my laptop so that the built-in camera is no longer suspended.  So long as it can be computed quickly, the list of cameras should be a computed property.  If it can't be computed quickly, it should be cached and re-computed after every access so that it will update after being computed.

2. Tracking the selected camera by its index, rather than a uniqueID, can cause errors if the set of cameras is allowed to change (as it should be.)

For example, if there are 4 cameras (0..3) and the selected camera is camera with index 2. If Camera 1 is disconnected, the selectedCameraIndex would represented what had previously been been the camera at index 3 instead of the correct camera (previously index 2 and now index 1.)

**Comment by UppaJung** [comment]
The selection box doesn't work.

[Screen Recording 2021-01-20 at 8.34.36 AM.zip](https://github.com/dicekeys/dicekeys-app-ios/files/5839213/Screen.Recording.2021-01-20.at.8.34.36.AM.zip)

File: DiceKeys/Views/Scanning/ScanDiceKey.swift [hunk 1]
```diff
@@ -21,11 +21,14 @@ struct ScanDiceKey: View {
...
         self.stickers = stickers
         self.onDiceKeyRead = onDiceKeyRead
+        getAvailableCameras()
     }
 
     @State var frameCount: Int = 0
     @State var facesRead: [FaceRead]?
     @State var processedImageFrameSize: CGSize?
+    @State var selectedCameraIndex: Int = 0
+    private var availableCameras = [AVCaptureDevice]()
```
Comment (UppaJung) [inline DiceKeys/Views/Scanning/ScanDiceKey.swift:31]: Non-state mutated properties are best avoided.

Why not make it a computed property?

Comment (Abhityagi16) [reply DiceKeys/Views/Scanning/ScanDiceKey.swift:31]: I prefer avoiding computing properties whenever I can. It is not an issue right now but as the app grows too many computed properties make their way and it becomes very slow.

Anyway, in this case I didn't think we had to update camera list in real time.

**Comment by Abhityagi16** [comment]
> @Abhityagi16
> I'll try to get this to run locally. For now based on reviewing the code...
> 
> Issues to fix:
> 
> 1. The set of cameras is determined once and never updated, so if I connect or disconnect a camera the picker will not get the updated list. Nor will it get updated if I open the screen of my laptop so that the built-in camera is no longer suspended.  So long as it can be computed quickly, the list of cameras should be a computed property.  If it can't be computed quickly, it should be cached and re-computed after every access so that it will update after being computed.
> 2. Tracking the selected camera by its index, rather than a uniqueID, can cause errors if the set of cameras is allowed to change (as it should be.)
> 
> For example, if there are 4 cameras (0..3) and the selected camera is camera with index 2. If Camera 1 is disconnected, the selectedCameraIndex would represented what had previously been been the camera at index 3 instead of the correct camera (previously index 2 and now index 1.)

1. Yes, the camera list updating is not real time. 
2. It's very tricky to use Picker in SwiftUI. Index is the only way I could get it to work. Initially I had used selectedCamera to be the AVCaptureDevice object but it wouldn't show up in selection. After digging around I found a stackoverflow answer which suggested using index and it worked. I'll try to dig further and see if we can somehow use uniqueId with Picker. SwiftUI does have some limitations and not enough documentation.

**Comment by UppaJung** [comment]
I figured out that the problem with the Picker in SwiftUI is that it doesn't work if the bound variable is allowed to be nil.

**Pull request merged**